src/panels/lovelace/editor/hui-dialog-edit-card.ts

```diff
@@ -84,17 +103,59 @@ export class HuiDialogEditCard extends LitElement {
     this._previewEl.yaml = ev.detail.yaml;
   }
 
+  private _handleConfigChanged(ev) {
+    console.log("Config Changed");
+
+    if (!this._previewEl) {
+      return;
+    }
+
+    this._previewEl.config = ev.detail.config;
+  }
+
   private _closeDialog() {
     this._dialog.close();
   }
 
+  private _toggleEditor() {
+    if (!this._elementConfig) {
+      return;
+    }
+    this._showUIEditor = !this._showUIEditor;
+    // This will center the dialog with the updated config
+    fireEvent(this._dialog, "iron-resize");
+  }
+
   private async _loadConfig() {
     this._cardConfig = await getCardConfig(this.hass!, this._cardId!);
+    this._loadElementConfig();
     await this.updateComplete;
     // This will center the dialog with the updated config
     fireEvent(this._dialog, "iron-resize");
   }
 
+  private async _loadElementConfig() {
+    const conf = yaml.safeLoad(this._cardConfig);
+    const elClass = customElements.get(`hui-${conf.type}-card`);
+
+    try {
+      this._elementConfig = await elClass.getElementConfig();
+    } catch (err) {
+      // No Element Config Function on Element
+      this._showUIEditor = false;
+      return;
+    }
+    if (!this._elementConfig) {
+      this._showUIEditor = false;
+      return;
+    }
+    this._elementConfig.setConfig(conf);
+    this._elementConfig.hass = this.hass;
+    this._elementConfig.addEventListener("config-changed", (ev) =>
+      this._handleConfigChanged(ev)
```

Comment:
handleConfigChanged should not be exposed to the implementation detail of how we got the config. I also think that we should keep track of the format.

```suggestion
      this._handleConfigChanged('js', ev.detail.config)
```

(for YAML, we would do `this._handleConfigChanged('yaml', yamlString)`

Reply:
For Yaml aren't we using fireEvent which gives the detail the yaml string? How can I give `hui-yaml-editor` on changed event the parameters? Unless we just target the element and add the change event listener as we do the form editor

Reply:
Wait…

`config-changed` and `yaml-changed` are both fired by the element using `fireEvent` to notify parent component that there is an updated event.

We either listen to this event by adding an event listener to the element using `addEventListener` or by specifying `@<event-name>="${this._handleX}"` as an attribute in a Lit template (which does exactly the same under the hood).

So how do we distinguish between the two events? Simple, 2 event listeners: `_handleYAMLChangedEvent` and `_handleConfigChangedEvent`.  Both of them will call a third method under the hood, example let's call it `_handleConfigChanged(format, value)`